Add a compilerError when trying to use the GMP module with CHPL_GMP=none

ronawho opened the pull request:
Currently, if you try to use the GMP module and CHPL_GMP=none, you'll get
non-obvious errors from the backend compiler about implicit declarations of GMP
types and GMP functions. This is especially confusing with speculative builds
of GMP, where the build failure is hidden and CHPL_GMP is silently set to none.
As an improvement, give an explicit compilerError that GMP support is not
available if CHPL_GMP=none.

This also updates a pre-existing error message in the regexp module to be
consistent with the wording used in the GMP one.

Note that the GMP error currently points to the GMP module instead of the user
line. This is a bug: https://github.com/chapel-lang/chapel/issues/8191

Related to https://github.com/chapel-lang/chapel/issues/7697

ronawho commented on test/library/standard/GMP/elliot/check-gmp-none-compiler-error.good:
```diff
@@ -0,0 +1 @@
+$CHPL_HOME/modules/standard/GMP.chpl:1226: error: GMP support not compiled in
```
The bad/wrong part of this PR is that we're reporting the internal GMP line number, not the user line number. I think this is because the compilerError is just top-level in the module, and not inside a user-callable function. I'm not sure I understand the compilers handling of error reporting to solve this myself though.

lydia-duncan replied:
That would probably require recognizing `use`s as contributing to the module init call chain, which is something our compiler doesn't do strongly right now.  If there's a way to remove the line number completely from the output, I would consider that a suitable compromise (though I'm not sure that's possible)

bradcray replied:
Does the Regexp module have the same issue?  compilerError() takes an errorDepth argument which should permit you to pop up a few levels of callchain potentially, but I can never quite remember how it works nor how well it works.  See test/param/diten/errorDepth.chpl, e.g.

ronawho replied:
Regexp has the compilerError inside the `compile()` proc, since you have to compile a regex in order to use it, so it correctly reports the user error line.  For gmp we _could_ make wrappers for all the gmp init routines and put the compilerError in there, but there's a lot of init routines (and currently they're all just naked externs)

I tried an errorDepth of 0, but that didn't change anything since I think that's the default

bradcray replied:
I'm OK with putting this in as-is and striving to improve it in the future.

ronawho commented on modules/standard/GMP.chpl:
```diff
@@ -1222,9 +1222,9 @@ module GMP {
-      expression and operator overloads.
-  */
+  if CHPL_GMP == "none" {
+    compilerError("GMP support not compiled in");
```
I'm open to different error messages -- this is just based on a similiar error message for regexp support: https://github.com/chapel-lang/chapel/blob/master/modules/standard/Regexp.chpl#L414

bradcray replied:
This error is better than no error, but I'd probably want something that helped a newbie user help themselves more (in both GMP and Regexp cases).  Referring directly to `CHPL_GMP` being set to `none` might help them find what they need with a docs search.  Or referring to the need to set CHPL_GMP to a value other than none and rebuilding might also be useful.

ronawho replied:
Do you like any of these?

 - "GMP support is not available with CHPL_GMP=none"
 - "Cannot use GMP with CHPL_GMP=none"
 - "Cannot use GMP with GMP_GMP=none, see https://chapel-lang.org/docs/master/usingchapel/chplenv.html#chpl-regexp for more information"


I don't love hardcoding a link, but I also don't love saying something like "set CHPL_GMP and rebuild" because it doesn't feel obvious that CHPL_GMP is an env variable, and it also doesn't seem like it's obvious that we mean rebuild the chpl runtime/third-party vs "recompile" your app

lydia-duncan replied:
Heads up, you're linking against chpl-regexp, not chpl-gmp

bradcray replied:
Maybe the second one?  I could also imagine changing it to say "the GMP module" rather than just "GMP".

It occurs to me belatedly that we should probably add a similar error in the BigInteger module.

lydia-duncan replied:
Wouldn't the BigInteger module benefit from the GMP error, since it uses the GMP module?

lydia-duncan replied:
Or do you mean that the error should be more explicit, since it isn't necessarily clear that BigInteger relies on GMP?

bradcray replied:
Exactly.  And I was thinking / hoping that putting a more explicit error into BigInteger.chpl would cause it to print first (though now I'm worrying that that's incorrect...).

ronawho replied:
Since BigInteger uses GMP you get the compilerError from the gmp module, and I couldn't emit a different error message for BigInteger, even if I put the compilerError before the [`use GMP;`](https://github.com/chapel-lang/chapel/blob/master/modules/standard/BigInteger.chpl#L112).

That's part of the reason I was trying to leave the error a little more vague and not say "the gmp module"

bradcray replied:
I see.  OK, let's go with this then and wait for users to express confusion before trying to do something more involved.

ronawho commented:
Alright, pushed a slightly updated error message (and made the regexp one match).

I also converted the test to a future (so changes to GMP.chpl line numbers won't cause testing failures) and added a smaller reproducer for the issue.

Any other comments or good to merge?

bradcray commented:
I haven't given it a line-by-line review or anything like that (esp the tests) but what I've seen is above the bar and I'm guessing the rest of it is reasonably straightforward.

ronawho commented:
I'm not sure that error would make much sense if a user is just doing a `use BigInteger;`, so I'll probably leave this alone for now
